### PR TITLE
add pool-agent

### PR DESCRIPTION
**Pull request by gamoutatsumi**
ref: #22 #23

### CODE REVIEW

File: pool-agent/cmd/agent.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+func generateInstanceName() (string, error) {
+	var b [4]byte
+	_, err := rand.Read(b[:])
+	if err != nil {
+		return "", fmt.Errorf("generate random id: %w", err)
+	}
+	return fmt.Sprintf("myshoes-runner-%x", b), nil
+}
+
+func (a *Agent) checkInstances() error {
```
Comment (whywaita) [inline pool-agent/cmd/agent.go:175]: Please write comments that introduce `checkInstances` function.

- What is the check?
- Would it like to operate an instance?

Maybe you need to rename this function if no comments.

File: pool-agent/cmd/config.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	var s Config
+	if err := toml.Unmarshal(f, &s); err != nil {
+		return Config{}, fmt.Errorf("parse config file: %w", err)
+	}
+	return s, nil
+}
+
+// ParseImageAlias parses LXD image alias.
+func ParseImageAlias(s string) (api.InstanceSource, error) {
```
Comment (whywaita) [inline pool-agent/cmd/config.go:46]: `ParseImageAlias` is the same function in shoes-lxd-multi-server.
https://github.com/whywaita/shoes-lxd-multi/blob/58c0274362ad68916b3afcc2015a6f8e08ac69e0/server/pkg/api/server_add_instance.go#L282

Could you update to export function in shoes-lxd-multi-server?
We won't maintenanse a same code in other functions.

File: pool-agent/cmd/create.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	"log/slog"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/lxc/lxd/shared/api"
+)
+
+func (a *Agent) createInstance(name string, rt ResourceTypesMap, l *slog.Logger) error {
+	l.Info("Creating instance")
```
Comment (whywaita) [inline pool-agent/cmd/create.go:14]: Please set the instance name in the logger for operation.

File: server/pkg/api/pool.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+func findInstances(ctx context.Context, targets []lxdclient.LXDHost, match func(api.Instance) bool, limitOverCommit uint64, l *slog.Logger) []instance {
+	type result struct {
+		host              *lxdclient.LXDHost
+		overCommitPercent uint64
+		instances         []string
+	}
+	rs := make([]result, len(targets))
+
+	wg := new(sync.WaitGroup)
+	wg.Add(len(targets))
```
Comment (whywaita) [inline server/pkg/api/pool.go:103]: Could you move to inside of range? ( to `wg.Add(1)` )
Maybe this occurs in infinite waiting if failure.

File: server/pkg/api/server_add_instance.go [hunk 5]
```diff
@@ -230,7 +307,7 @@ func getResources(ctx context.Context, targetLXDHosts []lxdclient.LXDHost) ([]ta
 	for _, t := range targetLXDHosts {
 		t := t
 		eg.Go(func() error {
-			l := slog.With("host", t.HostConfig.LxdHost)
+			l := slog.With("stadium", t.HostConfig.LxdHost)
```
Comment (whywaita) [inline server/pkg/api/server_add_instance.go:310]: Please update all param if change host to stadium.

File: pool-agent/cmd/agent.go [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+var (
+	errAlreadyStopped = "The instance is already stopped"
+)
+
+type instances map[string]struct{}
+
+func newAgent(ctx context.Context) (*Agent, error) {
+	conf, err := LoadConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load config")
```
[the remark this answers is not in the document]
Comment (gamoutatsumi) [reply pool-agent/cmd/agent.go:48]: For no particular reason.
I probably didn't consider uniformity.
They will be fixed...

File: pool-agent/cmd/create.go (hunk 3 above)
Comment (gamoutatsumi) [reply pool-agent/cmd/create.go:14]: Instance name has already set by caller.
https://github.com/gamoutatsumi/shoes-lxd-multi/blob/39cb6bfe6a08ef177c2aafd4f9ee3742545a43a7/pool-agent/cmd/agent.go#L206

**Pull request merged**